File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.php
class Solution {

    /**
     * @param Integer[] $prices
     * @return Integer
     */
    function maxProfit($prices) {
        $n = sizeof($prices);
        $help = array();
        for($i=0; $i<$n; ++$i){
            $help[$prices[$i]]++;
        }
        $ans = 0;
        ksort($help);
        $help_key = array_keys($help);
        $cur = $n-1;
        for($i=0; $i<$n; ++$i){
            $help[$prices[$i]]--;
            while($cur>= 0 and $help[$help_key[$cur]] == 0)
                $cur--;
            $ans = max($ans, $help_key[$cur] - $prices[$i]);
        }
        return $ans;
    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
import java.util.*;

class Solution {

    public int maxProfit(int[] prices) {
        int n = prices.length;
        Map<Integer, Integer> help = new HashMap<>();
        
        for (int i = 0; i < n; ++i) {
            help.put(prices[i], help.getOrDefault(prices[i], 0) + 1);
        }
        
        int ans = 0;
        Integer[] helpKeys = help.keySet().toArray(new Integer[0]);
        Arrays.sort(helpKeys);
        
        int cur = helpKeys.length - 1;
        for (int i = 0; i < n; ++i) {
            help.put(prices[i], help.get(prices[i]) - 1);
            while (cur >= 0 && help.get(helpKeys[cur]) == 0) {
                cur--;
            }
            if (cur >= 0) {
                ans = Math.max(ans, helpKeys[cur] - prices[i]);
            }
        }
        
        return ans;
    }
}